src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/ItemMapper.java

```diff
@@ -1,103 +1,35 @@
-     *
-     * Used for adding context to GetItemRequest, PutItemRequest, UpdateItemRequest, or DeleteItemRequest objects.
+    /**
+     * Takes a virtual table record (not qualified with multitenant context) to be upserted and converts it into the
+     * corresponding physical table record (qualified with multitenant context). Used for adding context to
+     * put/update requests.
      */
-    Map<String, AttributeValue> apply(Map<String, AttributeValue> unqualifiedItem) {
-        Map<String, AttributeValue> qualifiedItem = new HashMap<>();
-        unqualifiedItem.forEach((field, attribute) -> {
-            List<FieldMapping> fieldMappings = virtualToPhysicalFieldMappings.get(field);
-            if (!isNullOrEmpty(fieldMappings)) {
-                fieldMappings.forEach(fieldMapping -> qualifiedItem.put(fieldMapping.getTarget().getName(),
-                    fieldMapping.isContextAware()
-                        ? fieldMapper.apply(fieldMapping, attribute)
-                        : attribute));
-            } else {
-                qualifiedItem.put(field, attribute);
-            }
-        });
-
-        return qualifiedItem;
-    }
+    Map<String, AttributeValue> applyForWrite(Map<String, AttributeValue> virtualItem);
 
-    /*
-     * Takes a map representing a record in a physical table that is effectively qualified with multitenant context and
-     * returns a map representing a record in the virtual table with qualifications removed.
-     *
-     * Used for removing context from GetItemResult, QueryResult, or ScanResult objects.
+    /**
+     * Takes a virtual table record (not qualified with multitenant context) and returns the corresponding physical
+     * table record with the table primary key attribute(s) only, or if a virtual secondary index is specified, with the
+     * table table primary key attribute(s) and the secondary index attribute(s) only.
      */
-    Map<String, AttributeValue> reverse(Map<String, AttributeValue> qualifiedItem) {
-        if (qualifiedItem == null) {
-            return null;
-        }
-        Map<String, AttributeValue> unqualifiedItem = new HashMap<>();
-        qualifiedItem.forEach((field, attribute) -> {
-            List<FieldMapping> fieldMappings = physicalToVirtualFieldMappings.get((field));
-            if (fieldMappings != null && !fieldMappings.isEmpty()) {
-                fieldMappings.forEach(fieldMapping -> unqualifiedItem.put(fieldMapping.getTarget().getName(),
-                    fieldMapping.isContextAware()
-                        ? fieldMapper.reverse(fieldMapping, attribute)
-                        : attribute));
-            } else {
-                unqualifiedItem.put(field, attribute);
-            }
-        });
-        return unqualifiedItem;
-    }
+    Map<String, AttributeValue> applyToKeyAttributes(Map<String, AttributeValue> virtualItem,
+                                                     @Nullable DynamoSecondaryIndex virtualSecondaryIndex);
```

Comment:
This is used when we are referring to an item by its table primary key only, or by a secondary index primary key if a non-null index argument is provided. Almost all uses of this are by table primary key only -- for specifying a record for update, delete, get, and batch get. The secondary index is specified only when we query or scan on an index, more specifically for translating a virtual ExclusiveStartKey into a physical one -- in this case, the start key should include all table PK attributes as well as the secondary index's PK attributes.